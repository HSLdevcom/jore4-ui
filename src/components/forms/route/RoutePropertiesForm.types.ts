import { z } from 'zod';
import { RouteDirection } from '../../../types/RouteDirection';
import {
  FormState as ConfirmSaveFormState,
  schema as confirmSaveFormSchema,
} from '../common/ConfirmSaveForm';

export const routeFormSchema = z
  .object({
    label: z.string().min(1),
    finnishName: z.string().min(1),
    on_line_id: z.string().uuid().min(1),
    direction: z.nativeEnum(RouteDirection),
  })
  .merge(confirmSaveFormSchema);

export type RouteFormState = z.infer<typeof routeFormSchema> &
  ConfirmSaveFormState;
