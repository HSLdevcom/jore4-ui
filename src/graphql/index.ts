export * from './auth';
export * from './client';
export * from './infrastructureNetwork';
// export all values from journey patterns when it has named exports
// eslint-disable-next-line import/export
export * from './journeyPattern';
export * from './route';
export * from './servicePattern';
export * from './types';
