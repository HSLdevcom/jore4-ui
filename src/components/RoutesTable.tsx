import React from 'react';
import { RouteRoute } from '../generated/graphql';
import { RoutesTableRow } from './RoutesTableRow'; // eslint-disable-line import/no-cycle

interface Props {
  className?: string;
  routes: RouteRoute[];
  testId?: string;
}

export const RoutesTable = ({
  className,
  routes,
  testId,
}: Props): JSX.Element => {
  return (
    <table className={`w-full ${className}`} data-testid={testId}>
      <tbody>
        {routes.map((item: RouteRoute) => (
          <RoutesTableRow key={item.route_id} route={item} />
        ))}
      </tbody>
    </table>
  );
};
