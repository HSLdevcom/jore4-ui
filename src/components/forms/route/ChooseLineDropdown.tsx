import React from 'react';
import { useTranslation } from 'react-i18next';
import { RouteLine, useListAllLinesQuery } from '../../../generated/graphql';
import {
  FormInputProps as ListboxInputProps,
  Listbox,
} from '../../../uiComponents';

interface Props extends ListboxInputProps {
  testId?: string;
}

const mapToOptionContent = (item: RouteLine) => (
  <span>{`${item.label} (${item.name_i18n})`}</span>
);

const mapToOption = (item: RouteLine) => ({
  key: item.line_id,
  value: item.line_id,
  render: () => mapToOptionContent(item),
});

export const ChooseLineDropdown = ({
  testId,
  value,
  onChange,
  onBlur,
}: Props): JSX.Element => {
  const { data: lines } = useListAllLinesQuery();
  const { t } = useTranslation();

  // @ts-expect-error typings are off?
  const options = lines?.route_line.map(mapToOption) || [];

  const selectedLine = lines?.route_line.find(
    (item) => item.line_id === value,
  ) as RouteLine;

  // if no line is selected, show "Choose line"
  const mapToButtonContent = (displayedLine?: RouteLine) => (
    <div className="w-full">
      {displayedLine
        ? `${displayedLine.label} (${displayedLine.name_i18n})`
        : t('routes.chooseLine')}
    </div>
  );

  return (
    <Listbox
      testId={testId}
      buttonContent={mapToButtonContent(selectedLine)}
      options={options}
      value={value}
      onChange={onChange}
      onBlur={onBlur}
    />
  );
};
