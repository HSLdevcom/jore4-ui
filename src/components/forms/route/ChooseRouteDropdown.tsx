import { DateTime } from 'luxon';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import {
  RouteRoute,
  useGetRouteDetailsByLabelWildcardQuery,
} from '../../../generated/graphql';
import { mapRoutesDetailsResult } from '../../../graphql';
import { MAX_DATE, MIN_DATE } from '../../../time';
import { Priority } from '../../../types/Priority';
import { Combobox, ComboboxInputProps } from '../../../uiComponents';
import { mapToVariables } from '../../../utils';
import { DateRange } from '../common/DateRange';

interface Props extends ComboboxInputProps {
  testId?: string;
  date: DateTime;
  priorities: Priority[];
}

const mapToOptionContent = (item: RouteRoute) => (
  <div className="flex flex-col">
    <div>
      <span className="font-bold">{item.label}</span>
      {` | ${item.description_i18n}`}
    </div>
    <div className="text-sm">
      <DateRange
        startDate={item.validity_start || MIN_DATE}
        endDate={item.validity_end || MAX_DATE}
      />
    </div>
  </div>
);

const mapToOption = (item: RouteRoute) => ({
  key: item.route_id,
  value: item.route_id,
  render: () => mapToOptionContent(item),
});

export const ChooseRouteDropdown = ({
  testId,
  value,
  onChange,
  onBlur,
  date,
  priorities,
}: Props): JSX.Element => {
  const { t } = useTranslation();

  const [query, setQuery] = useState('');

  const routesResult = useGetRouteDetailsByLabelWildcardQuery(
    mapToVariables({
      label: `${query}%`,
      date: date.toISO(),
      priorities,
    }),
  );

  const routes = mapRoutesDetailsResult(routesResult);

  const options = routes?.map(mapToOption) || [];

  const selectedRoute = routes?.find(
    (item) => item.route_id === value,
  ) as RouteRoute;

  // If no route is selected, show "Choose route"
  const mapToButtonContent = (displayedRoute?: RouteRoute) => {
    // Headless UI Combobox component takes care of displaying query
    if (query && !displayedRoute) return null;

    return (
      <div className="w-full">
        {displayedRoute
          ? `${displayedRoute.label} (${displayedRoute.description_i18n})`
          : t('routes.chooseRoute')}
      </div>
    );
  };

  return (
    <Combobox
      id="choose-route-combobox"
      testId={testId}
      buttonContent={mapToButtonContent(selectedRoute)}
      options={options}
      value={value}
      onChange={onChange}
      onBlur={onBlur}
      onQueryChange={setQuery}
    />
  );
};
