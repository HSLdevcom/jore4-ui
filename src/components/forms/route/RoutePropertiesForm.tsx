import { Switch as HuiSwitch } from '@headlessui/react';
import { zodResolver } from '@hookform/resolvers/zod';
import React, { useContext, useState } from 'react';
import { Controller, FormProvider, useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { MapEditorContext } from '../../../context/MapEditor';
import { Column, Row } from '../../../layoutComponents';
import { Switch, SwitchLabel } from '../../../uiComponents';
import { ConfirmSaveForm } from '../common/ConfirmSaveForm';
import { ChooseLineDropdown } from './ChooseLineDropdown';
import { DirectionDropdown } from './DirectionDropdown';
import { routeFormSchema, RouteFormState } from './RoutePropertiesForm.types';
import { TemplateRouteSelector } from './TemplateRouteSelector';

export interface RouteFormProps {
  id?: string;
  routeLabel?: string | null;
  className?: string;
  defaultValues: Partial<RouteFormState>;
  onSubmit: (state: RouteFormState) => void;
}

const RoutePropertiesFormComponent = (
  { id, routeLabel, className, defaultValues, onSubmit }: RouteFormProps,
  ref: ExplicitAny,
): JSX.Element => {
  const { t } = useTranslation();
  const {
    state: { editedRouteData, creatingNewRoute },
    dispatch,
  } = useContext(MapEditorContext);

  const methods = useForm<RouteFormState>({
    defaultValues,
    resolver: zodResolver(routeFormSchema),
  });

  const [showTemplateRouteSelector, setShowTemplateRouteSelector] =
    useState(false);

  const {
    register,
    handleSubmit,
    formState: { errors },
    control,
  } = methods;

  const setTemplateRoute = (uuid?: UUID) => {
    dispatch({
      type: 'setState',
      payload: {
        editedRouteData: {
          ...editedRouteData,
          templateRouteId: uuid,
        },
      },
    });
  };

  return (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <FormProvider {...methods}>
      <form
        id={id || 'route-properties-form'}
        className={className || ''}
        onSubmit={handleSubmit(onSubmit)}
        ref={ref}
      >
        {routeLabel && (
          <Row>
            <h2 className="mb-8 text-2xl font-bold">
              {t('routes.route')} {routeLabel}
            </h2>
          </Row>
        )}
        <Row className="mb-5 flex-wrap gap-2">
          <Column className="w-80 flex-auto">
            <label htmlFor="description_i18n">{t('routes.name')}</label>
            <input
              id="description_i18n"
              type="text"
              {...register('description_i18n', {})}
            />
            <p>
              {errors.description_i18n?.type === 'too_small' &&
                t('formValidation.required')}
            </p>
          </Column>
          <Column className="w-44 flex-auto">
            <label htmlFor="label">{t('routes.label')}</label>
            <input id="label" type="text" {...register('label', {})} />
            <p>
              {errors.label?.type === 'too_small' &&
                t('formValidation.required')}
            </p>
          </Column>
          <Column className="w-44 flex-auto">
            <label htmlFor="direction">{t('routes.direction')}</label>
            <Controller
              name="direction"
              control={control}
              render={({ field: { onChange, onBlur, value } }) => (
                <DirectionDropdown
                  value={value}
                  onChange={onChange}
                  onBlur={onBlur}
                />
              )}
            />
            <p>
              {errors.direction?.type === 'invalid_enum_value' &&
                t('formValidation.required')}
            </p>
          </Column>
          <Column className="w-80 flex-auto">
            <label htmlFor="on_line_id">{t('routes.addToLine')}</label>
            <Controller
              name="on_line_id"
              control={control}
              render={({ field: { onChange, onBlur, value } }) => (
                <ChooseLineDropdown
                  value={value}
                  onChange={onChange}
                  onBlur={onBlur}
                />
              )}
            />
            <p>
              {errors.on_line_id?.type === 'invalid_type' &&
                t('formValidation.required')}
            </p>
          </Column>
          {creatingNewRoute && (
            <>
              <Row className="flex-auto items-center">
                <HuiSwitch.Group>
                  <SwitchLabel className="my-1 mr-2">
                    {t('routes.useTemplateRoute')}
                  </SwitchLabel>
                  <Switch
                    checked={showTemplateRouteSelector}
                    onChange={(enabled: boolean) => {
                      setShowTemplateRouteSelector(enabled);

                      if (!enabled) {
                        setTemplateRoute(undefined);
                      }
                    }}
                  />
                </HuiSwitch.Group>
              </Row>
              {showTemplateRouteSelector && (
                <TemplateRouteSelector
                  value={editedRouteData.templateRouteId}
                  onChange={(e) => setTemplateRoute(e.target.value)}
                />
              )}
            </>
          )}
        </Row>
        <Row className="mt-7 border-t">
          <ConfirmSaveForm className="mt-5" />
        </Row>
      </form>
    </FormProvider>
  );
};

export const RoutePropertiesForm = React.forwardRef(
  RoutePropertiesFormComponent,
);
