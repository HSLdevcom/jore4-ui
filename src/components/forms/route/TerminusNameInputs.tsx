import React from 'react';
import { useTranslation } from 'react-i18next';
import { Column, Row } from '../../../layoutComponents';
import { StringInput } from '../common/StringInput';
import { RouteFormState } from './RoutePropertiesForm.types';

const testIds = {
  originFinnishNameInput: 'terminusNamesForm:origin:finnishNameInput',
  originFinnishShortNameInput: 'terminusNamesForm:origin:finnishShortNameInput',
  originSwedishNameInput: 'terminusNamesForm:origin:swedishNameInput',
  originSwedishShortNameInput: 'terminusNamesForm:origin:swedishShortNameInput',
  destinationFinnishNameInput: 'terminusNamesForm:destination:finnishNameInput',
  destinationFinnishShortNameInput:
    'terminusNamesForm:destination:finnishShortNameInput',
  destinationSwedishNameInput: 'terminusNamesForm:destination:swedishNameInput',
  destinationSwedishShortNameInput:
    'terminusNamesForm:destination:swedishShortNameInput',
};

interface Props {
  className?: string;
}

export const TerminusNameInputs = ({ className = '' }: Props): JSX.Element => {
  const { t } = useTranslation();
  return (
    <div className={`${className} w-full`}>
      <h2 className="w-full pt-1 text-xl font-bold">
        {t('routes.origin.title')}
      </h2>
      <Row className="mt-5 flex-wrap gap-2">
        <Column className="w-80 flex-auto">
          <StringInput<RouteFormState>
            translationPrefix="routes"
            fieldPath="origin.name.fi_FI"
            testId={testIds.originFinnishNameInput}
          />
        </Column>
        <Column className="w-80 flex-auto">
          <StringInput<RouteFormState>
            translationPrefix="routes"
            fieldPath="origin.shortName.fi_FI"
            testId={testIds.originFinnishShortNameInput}
          />
        </Column>
      </Row>
      <Row className="mt-5 flex-wrap gap-2">
        <Column className="w-80 flex-auto">
          <StringInput<RouteFormState>
            translationPrefix="routes"
            fieldPath="origin.name.sv_FI"
            testId={testIds.originSwedishNameInput}
          />
        </Column>
        <Column className="w-80 flex-auto">
          <StringInput<RouteFormState>
            translationPrefix="routes"
            fieldPath="origin.shortName.sv_FI"
            testId={testIds.originSwedishShortNameInput}
          />
        </Column>
      </Row>
      <h2 className="w-full pt-1 text-xl font-bold">
        {t('routes.destination.title')}
      </h2>
      <Row className="mt-5 flex-wrap gap-2">
        <Column className="w-80 flex-auto">
          <StringInput<RouteFormState>
            translationPrefix="routes"
            fieldPath="destination.name.fi_FI"
            testId={testIds.destinationFinnishNameInput}
          />
        </Column>
        <Column className="w-80 flex-auto">
          <StringInput<RouteFormState>
            translationPrefix="routes"
            fieldPath="destination.shortName.fi_FI"
            testId={testIds.destinationFinnishShortNameInput}
          />
        </Column>
      </Row>
      <Row className="mt-5 flex-wrap gap-2">
        <Column className="w-80 flex-auto">
          <StringInput<RouteFormState>
            translationPrefix="routes"
            fieldPath="destination.name.sv_FI"
            testId={testIds.destinationSwedishNameInput}
          />
        </Column>
        <Column className="w-80 flex-auto">
          <StringInput<RouteFormState>
            translationPrefix="routes"
            fieldPath="destination.shortName.sv_FI"
            testId={testIds.destinationSwedishShortNameInput}
          />
        </Column>
      </Row>
    </div>
  );
};
