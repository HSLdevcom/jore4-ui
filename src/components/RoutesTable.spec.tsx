import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { useListChangingRoutesQuery } from '../generated/graphql';
import { mapListChangingRoutesResult } from '../graphql/route';
import { render } from '../utils/test-utils';
import { RoutesTable } from './RoutesTable';

describe(`<${RoutesTable.name} />`, () => {
  const testId = 'routesTable1';
  const route1Id = '03d55414-e5cf-4cce-9faf-d86ccb7e5f98';
  const lineId = '101f800c-39ed-4d85-8ece-187cd9fe1c5e';
  const stop1Id = 'f8eace87-7901-4438-bfee-bb6f24f1c4c4';
  const stop2Id = 'e3528755-711f-4e4f-9461-7931a2c4bc6d';

  // this response is copy-pasted from the actual graphql response
  const mockResponse = {
    data: {
      route_route: [
        {
          route_id: route1Id,
          description_i18n: null,
          on_line_id: lineId,
          __typename: 'route_route',
          route_line: {
            line_id: lineId,
            label: '65',
            name_i18n: 'Rautatientori - Kätilöopisto - Veräjälaakso',
            short_name_i18n: 'Rautatientori - Veräjälaakso',
            description_i18n: 'Rautatientori - Kätilöopisto - Veräjälaakso',
            __typename: 'route_line',
          },
          starts_from_scheduled_stop_point: {
            scheduled_stop_point_id: stop1Id,
            label: 'pysäkki B',
            __typename: 'service_pattern_scheduled_stop_point',
          },
          ends_at_scheduled_stop_point: {
            scheduled_stop_point_id: stop2Id,
            label: 'pysäkki A',
            __typename: 'service_pattern_scheduled_stop_point',
          },
        },
      ],
    },
  } as ReturnType<typeof useListChangingRoutesQuery>;

  test('Renders the routes table based on data', async () => {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const routes = mapListChangingRoutesResult(mockResponse)!;
    const { asFragment } = render(
      <BrowserRouter>
        <RoutesTable testId={testId} routes={routes} />
      </BrowserRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
