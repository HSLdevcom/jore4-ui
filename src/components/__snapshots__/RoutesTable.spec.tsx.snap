// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoutesTable /> Renders the routes table based on data 1`] = `
<DocumentFragment>
  <table
    class="w-full undefined"
    data-testid="routesTable1"
  >
    <tbody>
      <tr
        class="border undefined"
      >
        <td
          class="border-l-8 border-hsl-dark-green py-4 pl-16 pr-4 font-bold"
        >
          <a
            href="/lines/101f800c-39ed-4d85-8ece-187cd9fe1c5e"
          >
            <div
              class="flex flex-row undefined"
            >
              <div
                class="flex flex-col w-1/2"
              >
                <p
                  class="text-3xl"
                />
                <p
                  class="text-lg"
                >
                  pysäkki B - pysäkki A
                </p>
              </div>
              <div
                class="flex flex-col w-1/2 text-right"
              >
                <p
                  class="text-lg font-bold"
                >
                  Voimassa  - 
                </p>
                <p
                  class="text-lg"
                >
                  !Muokattu dd.mm.yyyy hh:mm | S. Suunnittelija
                </p>
              </div>
            </div>
          </a>
        </td>
        <td
          class="border"
        >
          <a
            class="flex justify-center py-3"
            href="/map?routeId=03d55414-e5cf-4cce-9faf-d86ccb7e5f98"
          >
            <svg
              class="text-center text-5xl text-tweaked-brand"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M18 8c0-3.31-2.69-6-6-6S6 4.69 6 8c0 4.5 6 11 6 11s6-6.5 6-11zm-8 0c0-1.1.9-2 2-2s2 .9 2 2a2 2 0 11-4 0zM5 20v2h14v-2H5z"
              />
            </svg>
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
